core/src/main/java/org/renjin/primitives/S3.java

```diff
@@ -258,6 +268,314 @@ public static SEXP tryDispatchFromPrimitive(Context context, Environment rho, Fu
+      throw new EvalException("No S4 method found for '" + opName + "'");
+    }
+    
+    return methodEnvironment;
+  }
+  
+  /**
+   * Provided a PairList of arguments and the number of arguments used for signature
+   * this function first evaluates the arguments and extracts their class and superclass
+   * information.
+   *
+   * */
+  private static ArrayList<MethodRanking> generatePossibleSignatures(Context context, Environment rho,
+                                                                     PairList args, int depth) {
+  
+  
+    ArgumentSignature[] argSignatures = new ArgumentSignature[depth];
+  
+    for(int i = 0; i < depth; i++) {
+      String argumentClass = evaluateAndGetClass(context, args, rho, i);
+      argSignatures[i] = getAllClassAndDistance(context, argumentClass);
+    }
+  
+    int numberOfPossibleSignatures = 1;
+    for(int i = 0; i < argSignatures.length; i++) {
+      numberOfPossibleSignatures = numberOfPossibleSignatures * argSignatures[i].getArgument().length;
+    }
+  
+    ArrayList<MethodRanking> possibleSignatures = new ArrayList<>(numberOfPossibleSignatures);
+  
+    int numberOfClassesCurrentArgument;
+    int argumentClassIdx = 0;
+    int repeat = 1;
+    int repeatIdx = 1;
+  
+    for(int col = 0; col < depth; col++) {
+      numberOfClassesCurrentArgument = argSignatures[col].getArgument().length;
+      for(int row = 0; row < numberOfPossibleSignatures; row++, repeatIdx++) {
+        if(argumentClassIdx == numberOfClassesCurrentArgument) {
+          argumentClassIdx = 0;
+        }
+        ArgumentSignature argSignature = argSignatures[col];
+        String signature = argSignature.getArgument(argumentClassIdx);
+        if(possibleSignatures.isEmpty() || possibleSignatures.toArray().length < row + 1 || possibleSignatures.toArray()[row] == null) {
+          int[] distance = argSignature.getDistanceAsArray(argumentClassIdx);
+          possibleSignatures.add(row, new MethodRanking(signature, distance));
+        } else {
+          int distance = argSignature.getDistance(argumentClassIdx);
+          possibleSignatures.set(row, possibleSignatures.get(row).append(signature, distance));
+        }
+        if(repeat == 1) {
+          argumentClassIdx++;
+        }
+        if(repeat != 1 && repeatIdx == repeat) {
+          repeatIdx = 0;
+          argumentClassIdx++;
+        }
+      }
+      repeatIdx = 1;
+      argumentClassIdx = 0;
+      repeat = repeat * numberOfClassesCurrentArgument;
+    }
+    return possibleSignatures;
+  }
+  
+  private static String evaluateAndGetClass(Context context, PairList args, Environment rho, int argumentIndex) {
+    // To get the class of an argument we have to evaluate the argument first. If its an atomic, than it lacks class
+    // attribute. In this case we use Attributes.getClass() to get the class name.
+    String className;
+    SEXP evaluatedArg = context.evaluate(args.getElementAsSEXP(argumentIndex), rho);
```

Comment:
Since `evaluatedArg` is not included in this method's return value, then it seems only possible that the selected method will get an unevaluated copy of the argument and the argument may be evaluated twice.